### PR TITLE
Updated Princess Quest

### PR DESCRIPTION
**Pull request by SewerCzar**
Similar to before. Key that is baggable -> hex next to princess -> door opens.

### CODE REVIEW

**Review by Leolus — DISMISSED** [review]
Set that door attack value higher and have it smack people who try to open it without a key? lol

**Review by zrikzlok — DISMISSED** [review]

**Comment by zrikzlok** [comment]
> Set that door attack value higher and have it smack people who try to open it without a key? lol

That would be pretty funny

**Review by Chilly021 — CHANGES_REQUESTED** [review]
The code still needs to be cleaned up from the prior addition. Jailer and Scroll are still in the Segment portion. 

 public class TrollJailer : VendorEntity
    {
        public TrollJailer()
        {
            Name = "Near Sighted Jailer";
            Body = 148;
    
            Health = MaxHealth = 50;
            BaseDodge = 10;
    
            Alignment = Alignment.Neutral;
    
            Experience = 50;
    
            Movement = 0;
        }
    
        public TrollJailer(Serial serial) : base(serial)
        {
        }
    
        public override void HandleOrder(OrderEventArgs args)
        {
            base.HandleOrder(args);
    
            if (args.Handled)
                return;
    
            var source = args.Source;
            var order = args.Order;
    
            /* example: sell key */
            if (_craft.TryGetMatch(order, out var craftMatch) && craftMatch.Groups[1].Value.Matches("princess", true))
            {
                args.Handled = true;
    
                if (AtCounter(source, out var counter))
                {
                    var items = Segment.GetItemsAt(Location).ToList();
    
                    var jailScroll = items.OfType<PrincessScroll>();
    
                    if (jailScroll.Any())
                    {
                        if (ConsumeFromLocation<PrincessScroll>(counter, 1))
                        {
                            var tile = source.Segment.FindTile(5, 8, 10);
    
                            if (tile != null && tile.ContainsComponent<Obstruction>())
                            {
                                var component = tile.GetComponent<Obstruction>();
                                tile.Remove(component);
                            }
                            Say($"Order from king... good. You feed princess. I take break.");
    
                            Timer.DelayCall(TimeSpan.FromSeconds(0.2), () => Delete());
                        }
                    }
                    else
                    {
                        Say($"Me not do anyt'ing wi'fout orders.");
                    }
                }
                else
                {
                    Say($"Me not hear great. Closer.");
                }
    
                return;
            }
        }
    }
    
    public class PrincessScroll : ItemEntity
    {
        /// <inheritdoc />
        public override int LabelNumber => 6000078;

        /// <inheritdoc />
        public override uint BasePrice => 100;

        /// <inheritdoc />
        public override int Weight => 5;
        
        /// <summary>
        /// Initializes a new instance of the <see cref="PrincessScroll"/> class.
        /// </summary>
        public PrincessScroll() : base(128)
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="PrincessScroll"/> class.
        /// </summary>
        public PrincessScroll(Serial serial) : base(serial)
        {
        }

        /// <inheritdoc />
        public override void GetDescription(List<LocalizationEntry> entries)
        {
            entries.Add(new LocalizationEntry(6200000, 6200244)); /* [You are looking at] [a scroll entitled 'Everything is Not As It May Appear'.] */
        }

        /// <inheritdoc />
        public override void Serialize(BinaryWriter writer)
        {
            base.Serialize(writer);

            writer.Write((short)1); /* version */
        }

        /// <inheritdoc />
        public override void Deserialize(BinaryReader reader)
        {
            base.Deserialize(reader);

            var version = reader.ReadInt16();

            switch (version)
            {
                case 1:
                {
                    break;
                }
            }
        }
    }

File: Segments/Underkingdom.mapproj [hunk 1]
```diff
@@ -102660,9 +102755,28 @@
...
+    
+    source.Emote("opens with a creak as key is inserted and turned in the lock.");
+    
+    corruptstone.Delete();
+    
+    var tile = source.Segment.FindTile(5, 8, 10);    
+    if (tile != null && tile.ContainsComponent<Obstruction>())
+    {
+        var component = tile.GetComponent<Obstruction>();
+        tile.Remove(component);
```
Comment (sproatba) [inline Segments/Underkingdom.mapproj:102776]: What puts back the door?

**Review by sproatba — CHANGES_REQUESTED** [review]
looks pretty good, just some cleanup I think

File: Segments/Underkingdom.mapproj [hunk 2]
```diff
@@ -616,6 +616,65 @@
...
         }
     }
+    
+    public class PrincessKey : Gem
+    {
+        /// <inheritdoc />
+        public override int LabelNumber => 6000055;
+
+        /// <inheritdoc />
+        public override uint BasePrice => 5000;
```
[the remark this answers is not in the document]
Comment (SewerCzar) [reply Segments/Underkingdom.mapproj:626]: I think most items cost more than they sell for. Do balms sell for 15? :p I just know its supposed to cost 50k what it sells for never... seemed... like an issue

[the remark this answers is not in the document]
Comment (sproatba) [reply Segments/Underkingdom.mapproj:626]: In other versions, yes balms and recalls and the key sold for what you bought them for.  Pawn shop items cost 3x what they were worth.  People would buy keys then sell them in town to keep encumbrance down, and the key was one of the best components for a thaum snake in the og as it was one of the most expensive items in the game.   More a question really than anything.  Most people aren't going to sell them anyways, and the snakes don't require an item to power them up.

File: Segments/Underkingdom.mapproj [hunk 3]
```diff
@@ -616,6 +616,65 @@
             }
         }
     }
+    
+    public class PrincessKey : Gem
```
[the remark this answers is not in the document]
Comment (sproatba) [reply Segments/Underkingdom.mapproj:620]: yeah power troll key and princess key were the same item.  We don't have to keep it that way, but farming keys off power troll was a way to keep costs down.

**Comment by sproatba** [comment]
After giving this a second look, I think we need to figure out how to deal with the scroll.  Someone could have a scroll, what happens if we nuke the scroll code?   We do need to clean up the old logic for sure, but need to do so in a way that doesn't screw something up with the items.

**Comment by SewerCzar** [comment]
> After giving this a second look, I think we need to figure out how to deal with the scroll. Someone could have a scroll, what happens if we nuke the scroll code? We do need to clean up the old logic for sure, but need to do so in a way that doesn't screw something up with the items.

Thats actually a good question: what happens if we delete a normal item cs file? I assume it would be the same results. Would lux have to purge it from inventory manually?

Do we have any items that... hm whats the logic on cloth.

**Review by Chilly021 — APPROVED** [review]

**Review by SewerCzar — COMMENTED** [review]
I made a Jeweled key file (miscellaneous items - compare it to silk finery) back in December when I was filling out all the art work into placeholders. We have access to a permanent item so problems solved on that front. Everything here is done in a way that can be changed if we want when locked doors are accessible.

**Review by sproatba — APPROVED** [review]

**Review by zrikzlok — APPROVED** [review]

**Pull request merged**